mathesar_ui/src/App.svelte

```diff
@@ -6,6 +6,15 @@
   import Header from '@mathesar/header/Header.svelte';
   import { toast } from '@mathesar/stores/toast';
   import { confirmationController } from '@mathesar/stores/confirmation';
+  import { currentSchemaId } from '@mathesar/stores/schemas';
+  import { beginUpdatingUrlWhenSchemaChanges } from './utils/routing';
+
+  // This is a bit of a hack to deal with our routing still being a patchwork of
+  // declarative and imperative logic. Without this call, the URL will not
+  // reliably set the query params when the schema changes. It actually _will_
+  // set the query params _sometimes_, but we weren't able to figure out why the
+  // behavior is inconsistent.
+  beginUpdatingUrlWhenSchemaChanges(currentSchemaId);
```

Comment:
@pavish This is how I'm handling that finicky routing issue we spent some time discussing yesterday.

I played with it for a while after our call. There's definitely something fishing going on which _inconsistently_ causes the stores within `TabList` to update after tinro has performed client-side navigation. I tried to trace it down using the in-browser debugging tools but just ended up in a rat's nest of compiled Svelte code.

So this call to `beginUpdatingUrlWhenSchemaChanges` is my hacky work-around. I had to put it within a pretty top-level location in the app in order to avoid cyclical dependency linting errors.

I'm interested in eventually moving Mathesar to SvelteKit (once it's stable), in which case we'll need do re-think all the routing anyway. So I'm okay with some hack in the mean time.

Reply:
> I'm interested in eventually moving Mathesar to SvelteKit (once it's stable), in which case we'll need do re-think all the routing anyway. So I'm okay with some hack in the mean time.

I estimate this mean time to be _atleast_ a couple years. Even then I wouldn't want to switch to SvelteKit until it has an official python adapter. Currently, if we need to switch, we'll need a node js middle layer, which is just pointless for us. There is a SvelteKit static adapter but that'll essentially the same (if not less perfomant) than directly using Svelte with pre-rendered data as we do now. So, I doubt we'd ever move to SvelteKit.

I'll take a look at this issue. We need to find the cause and fix it.

Reply:
It seems to be an issue with Svelte, or maybe it was intended but I don't see the behaviour mentioned in the docs. I opened https://github.com/sveltejs/svelte/issues/7330 for this.

Reply:
I'll add this to my list of things to discuss during our next sync call. I totally understand the need to manipulate the query params imperatively. I still think it's worth wrapping up this topic a bit with some more discussion. For example, I'd like to understand from you (if you're able to explain it) why the routing behavior is _inconsistent_. I understand the upstream issue with Svelte. But I'd still like to better understand the cause of the _inconsistency_.

Reply:
> There is a SvelteKit static adapter but that'll essentially the same (if not less perfomant) than directly using Svelte with pre-rendered data as we do now.

During our sync call, I'm also curious to explore this point above more, because that definitely does not match not my understanding.

Reply:
> I'd like to understand from you (if you're able to explain it) why the routing behavior is inconsistent

The routing behaviour is not inconsistent, it works exactly as it is intended. The router is not at fault here. The upstream issue from svelte caused our url update method to get called which modified the query params.

Can you explain what exactly you find inconsistent?

Reply:
It has to do with `TabContainer` being created and destroyed, since it is wrapped within an `#if` block.

Assume you opened a table in `Schema 1`, which creates the `TabContainer` and `TabList` instance, which inturn updates the url. Then you open `Schema 2`, which contains no tables, so `TabContainer` gets destroyed. Note, we create a new `TabList` instance when a schema is opened for the first time, so the url update will happen when `Schema 2` is opened.

When we switch back to `Schema 1`, the `TabContainer` gets created again, which updates the url. If you then open a table in `Schema 2`, switching between schemas do not update the url.

I assume in your local environment, you mostly tested by opening tables in both schemas, so you noticed the url update happening ~5% of the time.

Reply:
Interesting. Ok it sounds like there's a good chance I was not being observant enough of these small details while testing, and likely the behavior was consistent but my test scenarios were inconsistent. Thanks for explaining!